eth-rpc/serverdispatch.nim

```diff
@@ -0,0 +1,295 @@
+  return none(RpcJsonErrorContainer)
+
+proc wrapReply*(id: JsonNode, value: JsonNode, error: JsonNode): JsonNode =
+  return %{"jsonrpc": %"2.0", "result": value, "error": error, "id": id}
+
+proc sendError*(client: StreamTransport, code: int, msg: string,
+                id: JsonNode, data: JsonNode = newJNull()) {.async.} =
+  ## Send error message to client
+  let error = %{"code": %(code), "message": %msg, "data": data}
+  ifDebug: echo "Send error json: ",
+                wrapReply(newJNull(), error, id).pretty & "\c\l"
+  var answer = $wrapReply(id, newJNull(), error)
+  answer.add("\c\l")
+  result = client.write(addr answer[0], len(answer))
+
+proc sendJsonError*(state: RpcJsonError, client: StreamTransport,
+                    id: JsonNode, data = newJNull()) {.async.} =
+  ## Send client response for invalid json state
+  let errMsgs = jsonErrorMessages[state]
+  await client.sendError(errMsgs[0], errMsgs[1], id, data)
+
+proc processMessage(server: RpcServer, client: StreamTransport,
+                    line: string) {.async.} =
+  var node, id: JsonNode
+  # set up node and/or flag errors
+  var jsonErrorState = checkJsonErrors(line, node)
+
+  if jsonErrorState.isSome:
+    let errState = jsonErrorState.get
+    if errState.err == rjeInvalidJson:
+      # id cannot be retrieved
+      id = newJNull()
+    else:
+      id = if "id" in node: node["id"] else: newJNull()
+    await errState.err.sendJsonError(client, id, %errState.msg)
+  else:
+    let methodName = node["method"].str
+    id = node["id"]
+
+    if not server.procs.hasKey(methodName):
+      await client.sendError(METHOD_NOT_FOUND, "Method not found", id,
+                             %(methodName & " is not a registered method."))
+    else:
+      let res = await server.procs[methodName](node["params"])
+      var answer = $wrapReply(id, res, newJNull())
+      answer.add("\c\l")
+      discard await client.write(cast[pointer](addr answer[0]), len(answer))
+
+proc processClient(server: StreamServer,
+                   client: StreamTransport, udata: pointer) {.async.} =
+
+  while true:
+    let line = await client.readLine(limit = 1024)
+    if len(line) == 0:
+      client.close()
+      break
+
+    ifDebug: echo "Process client: ", $client.remoteAddress, ":" & line
+    var rpcsrv = cast[RpcServer](udata)
+    var fut = processMessage(rpcsrv, client, line)
+    yield fut
+    if fut.failed:
+      let err = fut.readError
+      if err of RpcProcError:
+        let rpcerr = RpcProcError(err)
+        await client.sendError(rpcerr.code, rpcerr.msg, rpcerr.data)
+      else:
+        await client.sendError(SERVER_ERROR, "Error", %getCurrentExceptionMsg())
+
+proc newRpcServer*(address: TransportAddress): RpcServer =
+  ## Create new instance of RPC server.
+  result = RpcServer(procs: newTable[string, RpcProc]())
+  result.server = createStreamServer(address, processClient, {ReuseAddr},
+                                     udata = cast[pointer](result))
```

Comment:
It's unfortunate that such pretty high level api requires `cast[pointer]`. It makes it much harder to reason about safety.

Reply:
About what safety you talking about? Could you please be more precise and specific?

Reply:
Talking about the memory safety in general. From this code it is obvious that a GC-ed object is passed as an untraced pointer to somewhere. So I need to take special care to inspect the lifetime of such untraced reference. Moreover, proper lifetimes have to be maintained in the future, which is just another thing to remember.

Reply:
`cast` circumvents the type system (I presume this is what @yglukhov means), and is best kept in the internals of the source API.

In this case, the RPC API shouldn't need to know about how async is handling data, so the `cast` should be 'hidden' inside `asyndispatch` as an accessor such as a template such as (for example)

```nim
template write*(transport: StreamTransport, msg: string): untyped = transport.write(cast[pointer](addr msg[0]), len(msg))
```

This incurs no overhead, as the template is inlined, but keeps it so the RPC code doesn't need to know how `asyncdispatch` handles it's internal data.

Reply:
So in this case (reasoning about the pointer), how would this be fixed? Surely it must use a pointer internally, and my suggestion of hiding this in fact makes reasoning more difficult since you don't know there's a cast.

Seems like the only way @yglukhov to avoid the reasoning issue is to not use pointers at all.

Reply:
Will this be possible?
```nim
proc createStreamServer(adr, proc, flags, data: ref T): Server =
  GC_ref(data)
  createStreamServer(adr, proc, flags & {UnrefUdata}, data)
```
Then not forget to unref it where relevant.